src/software/physics/euclidean_to_wheel.cpp

```diff
@@ -76,3 +80,109 @@ EuclideanSpace_t EuclideanToWheel::getEuclideanVelocity(
+        delta_target_wheel_velocity.cwiseAbs().maxCoeff();
+
+    // Step 2: Compare max delta velocity against the calculated maximum
+    if (max_delta_target_wheel_velocity > allowable_delta_wheel_velocity)
+    {
+        // Step 3: If larger, scale down to allowable max
+        ramp_wheel_velocity =
+            (delta_target_wheel_velocity / max_delta_target_wheel_velocity) *
+                allowable_delta_wheel_velocity +
+            current_wheel_velocity;
+    }
+    else
+    {
+        // If smaller, go straight to target
+        ramp_wheel_velocity = target_wheel_velocity;
+    }
+
+    // find absolute max wheel velocity
+    auto max_ramp_wheel_velocity = ramp_wheel_velocity.cwiseAbs().maxCoeff();
+
+    // compare against max wheel velocity
+    if (max_ramp_wheel_velocity > max_allowable_wheel_velocity)
+    {
+        // if larger, scale down to max
+        ramp_wheel_velocity = (ramp_wheel_velocity / max_ramp_wheel_velocity) *
+                              max_allowable_wheel_velocity;
+    }
+
+    return ramp_wheel_velocity;
+}
+
+std::unique_ptr<TbotsProto::DirectControlPrimitive> EuclideanToWheel::rampWheelVelocity(
+    const std::pair<Vector, AngularVelocity> current_local_velocity,
+    TbotsProto::DirectControlPrimitive& target_velocity_primitive,
+    const double& time_to_ramp)
+{
+    EuclideanSpace_t target_euclidean_velocity = EuclideanSpace_t::Zero();
+
+    TbotsProto::MotorControl motor_control = target_velocity_primitive.motor_control();
+    if (motor_control.has_direct_per_wheel_control())
+    {
+        TbotsProto::MotorControl_DirectPerWheelControl direct_per_wheel =
+            motor_control.direct_per_wheel_control();
+        WheelSpace_t wheel_velocity = {
+            direct_per_wheel.front_left_wheel_velocity(),
+            direct_per_wheel.back_left_wheel_velocity(),
+            direct_per_wheel.front_right_wheel_velocity(),
+            direct_per_wheel.back_right_wheel_velocity(),
+        };
+        target_euclidean_velocity = getEuclideanVelocity(wheel_velocity);
+    }
+    else if (target_velocity_primitive.motor_control().has_direct_velocity_control())
+    {
+        TbotsProto::MotorControl_DirectVelocityControl direct_velocity =
+            motor_control.direct_velocity_control();
+        target_euclidean_velocity = {
+            -direct_velocity.velocity().y_component_meters(),
+            direct_velocity.velocity().x_component_meters(),
+            direct_velocity.angular_velocity().radians_per_second()};
+    }
+
+    EuclideanSpace_t current_euclidean_velocity = {
+        -current_local_velocity.first.y(), current_local_velocity.first.x(),
+        current_local_velocity.second.toRadians()};
+
+    WheelSpace_t current_four_wheel_velocity =
+        getWheelVelocity(current_euclidean_velocity);
+
+    WheelSpace_t ramped_four_wheel = rampWheelVelocity(
+        current_four_wheel_velocity, target_euclidean_velocity, time_to_ramp);
```

Comment:
Other than the current velocities (`current_four_wheel_velocity` and `current_euclidean_velocity`), this seems to be pretty much identical to the logic in motor service:
https://github.com/UBC-Thunderbots/Software/blob/6525989aa74988a655cba9f9595dd7e0ce36d30f/src/software/jetson_nano/services/motor.cpp#L389-L427
Could we refactor motor service to use a simplified version of this function as well? The logic for current wheel velocities can probably be handled by the caller (motor service and erforcesimulator can handle the logic). Note that motor service doesn't actually use feedback from the motor encoders to get the wheel velocities, it just ramps based on the last ramped velocity. This is important as it allows us ramp smoothly.
https://github.com/UBC-Thunderbots/Software/blob/6525989aa74988a655cba9f9595dd7e0ce36d30f/src/software/jetson_nano/services/motor.cpp#L430
To make the logic portable for the two usecases, we should just return the ramped velocities (as the name suggests), and the logic for creating the `DirectVelocityControl` proto can probably be moved out to erforce sim as well.

Reply:
this makes sense, I made the change. I added a new method to er_force_simulator.cpp to handle the conversion to a primitive, so now motor.cpp doesn't have that whole switch statement thing